File: app/src/main/java/com/tablet/hid/util/TouchMouseConfigSerializer.kt
package com.tablet.hid.util

import com.tablet.hid.model.ButtonZoneConfig
import com.tablet.hid.model.ClickBehavior
import com.tablet.hid.model.KeyboardMacroButtonConfig
import com.tablet.hid.model.MacroHostDefaults
import com.tablet.hid.model.MouseButton
import com.tablet.hid.model.TouchMode
import com.tablet.hid.model.TouchMouseConfig
import com.tablet.hid.model.TouchMouseSubRegionConfig
import com.tablet.hid.model.ZoneType
import org.json.JSONArray
import org.json.JSONObject

object TouchMouseConfigSerializer {

    fun toCanonicalJson(config: TouchMouseConfig): JSONObject {
        val root = JSONObject()
        root.put("mode", config.mode.name)
        root.put("sensitivity", config.sensitivity)
        root.put("scrollEnabled", config.scrollEnabled)
        root.put("invertScroll", config.invertScroll)
        root.put("sharedDynamicZone", config.sharedDynamicZone)
        root.put("sharedDynamic", JSONObject().apply {
            put("offsetX", config.sharedDynamicOffsetX.toDouble())
            put("offsetY", config.sharedDynamicOffsetY.toDouble())
            put("radius",  config.sharedDynamicRadius.toDouble())
        })
        root.put("leftButton",  buttonZoneToJson(config.leftButton))
        root.put("rightButton", buttonZoneToJson(config.rightButton))
        root.put("sniper", JSONObject().apply {
            put("enabled", config.sniperEnabled)
            put("zone", JSONObject().apply {
                put("left",   config.sniperLeft.toDouble())
                put("top",    config.sniperTop.toDouble())
                put("right",  config.sniperRight.toDouble())
                put("bottom", config.sniperBottom.toDouble())
            })
            put("divisor", config.sniperDivisor.toDouble())
        })
        root.put("macroHostDefaults", config.macroHostDefaults.name)
        root.put("macroButtons", macrosToJson(config.macroButtons))
        return root
    }

    fun fromCanonicalJson(json: JSONObject): TouchMouseConfig {
        val defaults = TouchMouseConfig()
        val sharedDynamic = json.optJSONObject("sharedDynamic")
        val leftButton    = json.optJSONObject("leftButton")
        val rightButton   = json.optJSONObject("rightButton")
        val sniper        = json.optJSONObject("sniper")
        val sniperZone    = sniper?.optJSONObject("zone")
        return TouchMouseConfig(
            mode          = enumValueOrDefault(json.optString("mode"), defaults.mode),
            sensitivity   = json.optInt("sensitivity", defaults.sensitivity),
            scrollEnabled = json.optBoolean("scrollEnabled", defaults.scrollEnabled),
            invertScroll  = json.optBoolean("invertScroll", defaults.invertScroll),
            sharedDynamicZone = json.optBoolean("sharedDynamicZone", defaults.sharedDynamicZone),
            sharedDynamicOffsetX = sharedDynamic?.optDouble("offsetX", defaults.sharedDynamicOffsetX.toDouble())?.toFloat()
                                   ?: defaults.sharedDynamicOffsetX,
            sharedDynamicOffsetY = sharedDynamic?.optDouble("offsetY", defaults.sharedDynamicOffsetY.toDouble())?.toFloat()
                                   ?: defaults.sharedDynamicOffsetY,
            sharedDynamicRadius  = sharedDynamic?.optDouble("radius", defaults.sharedDynamicRadius.toDouble())?.toFloat()
                                   ?: defaults.sharedDynamicRadius,
            leftButton  = buttonZoneFromJson(leftButton, defaults.leftButton),
            rightButton = buttonZoneFromJson(rightButton, defaults.rightButton),
            sniperEnabled = sniper?.optBoolean("enabled", defaults.sniperEnabled) ?: defaults.sniperEnabled,
            sniperLeft    = sniperZone?.optDouble("left",   defaults.sniperLeft.toDouble())?.toFloat()   ?: defaults.sniperLeft,
            sniperTop     = sniperZone?.optDouble("top",    defaults.sniperTop.toDouble())?.toFloat()    ?: defaults.sniperTop,
            sniperRight   = sniperZone?.optDouble("right",  defaults.sniperRight.toDouble())?.toFloat()  ?: defaults.sniperRight,
            sniperBottom  = sniperZone?.optDouble("bottom", defaults.sniperBottom.toDouble())?.toFloat() ?: defaults.sniperBottom,
            sniperDivisor = sniper?.optDouble("divisor", defaults.sniperDivisor.toDouble())?.toFloat()   ?: defaults.sniperDivisor,
            macroHostDefaults = enumValueOrDefault(
                json.optString("macroHostDefaults"),
                defaults.macroHostDefaults,
            ),
            macroButtons = macrosFromJson(json.optJSONArray("macroButtons")),
        )
    }

    private fun buttonZoneToJson(btn: ButtonZoneConfig): JSONObject {
        val j = JSONObject()
        j.put("enabled", btn.enabled)
        j.put("zoneType", btn.zoneType.name)
        j.put("behavior", btn.behavior.name)
        j.put("staticZone", JSONObject().apply {
            put("left",   btn.staticLeft.toDouble())
            put("top",    btn.staticTop.toDouble())
            put("right",  btn.staticRight.toDouble())
            put("bottom", btn.staticBottom.toDouble())
        })
        j.put("dynamicZone", JSONObject().apply {
            put("offsetX", btn.dynamicOffsetX.toDouble())
            put("offsetY", btn.dynamicOffsetY.toDouble())
            put("radius",  btn.dynamicRadius.toDouble())
        })
        val subArr = JSONArray()
        btn.subRegions.forEach { subArr.put(subRegionToJson(it)) }
        j.put("subRegions", subArr)
        return j
    }

    private fun buttonZoneFromJson(j: JSONObject?, defaults: ButtonZoneConfig): ButtonZoneConfig {
        if (j == null) return defaults
        val staticZone  = j.optJSONObject("staticZone")
        val dynamicZone = j.optJSONObject("dynamicZone")
        val subArr      = j.optJSONArray("subRegions")
        return ButtonZoneConfig(
            enabled        = j.optBoolean("enabled", defaults.enabled),
            zoneType       = enumValueOrDefault(j.optString("zoneType"), defaults.zoneType),
            behavior       = enumValueOrDefault(j.optString("behavior"), defaults.behavior),
            staticLeft     = staticZone?.optDouble("left",   defaults.staticLeft.toDouble())?.toFloat()   ?: defaults.staticLeft,
            staticTop      = staticZone?.optDouble("top",    defaults.staticTop.toDouble())?.toFloat()    ?: defaults.staticTop,
            staticRight    = staticZone?.optDouble("right",  defaults.staticRight.toDouble())?.toFloat()  ?: defaults.staticRight,
            staticBottom   = staticZone?.optDouble("bottom", defaults.staticBottom.toDouble())?.toFloat() ?: defaults.staticBottom,
            dynamicOffsetX = dynamicZone?.optDouble("offsetX", defaults.dynamicOffsetX.toDouble())?.toFloat() ?: defaults.dynamicOffsetX,
            dynamicOffsetY = dynamicZone?.optDouble("offsetY", defaults.dynamicOffsetY.toDouble())?.toFloat() ?: defaults.dynamicOffsetY,
            dynamicRadius  = dynamicZone?.optDouble("radius",  defaults.dynamicRadius.toDouble())?.toFloat()  ?: defaults.dynamicRadius,
            subRegions     = if (subArr != null) List(subArr.length()) { i -> subRegionFromJson(subArr.getJSONObject(i)) } else defaults.subRegions,
        )
    }

    private fun subRegionToJson(sub: TouchMouseSubRegionConfig): JSONObject {
        val j = JSONObject()
        j.put("enabled", sub.enabled)
        j.put("zoneType", sub.zoneType.name)
        j.put("staticZone", JSONObject().apply {
            put("left",   sub.staticLeft.toDouble())
            put("top",    sub.staticTop.toDouble())
            put("right",  sub.staticRight.toDouble())
            put("bottom", sub.staticBottom.toDouble())
        })
        j.put("dynamicZone", JSONObject().apply {
            put("offsetX", sub.dynamicOffsetX.toDouble())
            put("offsetY", sub.dynamicOffsetY.toDouble())
            put("radius",  sub.dynamicRadius.toDouble())
        })
        j.put("keyboardModifiers", sub.keyboardModifiers)
        if (sub.alternateMouseButton != null) {
            j.put("alternateMouseButton", sub.alternateMouseButton.name)
        } else {
            j.put("alternateMouseButton", JSONObject.NULL)
        }
        return j
    }

    private fun subRegionFromJson(j: JSONObject): TouchMouseSubRegionConfig {
        val defaults    = TouchMouseSubRegionConfig()
        val staticZone  = j.optJSONObject("staticZone")
        val dynamicZone = j.optJSONObject("dynamicZone")
        val altBtn      = j.optString("alternateMouseButton").takeIf { it.isNotEmpty() }
        return TouchMouseSubRegionConfig(
            enabled  = j.optBoolean("enabled", defaults.enabled),
            zoneType = enumValueOrDefault(j.optString("zoneType"), defaults.zoneType),
            staticLeft     = staticZone?.optDouble("left",   defaults.staticLeft.toDouble())?.toFloat()   ?: defaults.staticLeft,
            staticTop      = staticZone?.optDouble("top",    defaults.staticTop.toDouble())?.toFloat()    ?: defaults.staticTop,
            staticRight    = staticZone?.optDouble("right",  defaults.staticRight.toDouble())?.toFloat()  ?: defaults.staticRight,
            staticBottom   = staticZone?.optDouble("bottom", defaults.staticBottom.toDouble())?.toFloat() ?: defaults.staticBottom,
            dynamicOffsetX = dynamicZone?.optDouble("offsetX", defaults.dynamicOffsetX.toDouble())?.toFloat() ?: defaults.dynamicOffsetX,
            dynamicOffsetY = dynamicZone?.optDouble("offsetY", defaults.dynamicOffsetY.toDouble())?.toFloat() ?: defaults.dynamicOffsetY,
            dynamicRadius  = dynamicZone?.optDouble("radius",  defaults.dynamicRadius.toDouble())?.toFloat()  ?: defaults.dynamicRadius,
            keyboardModifiers = j.optInt("keyboardModifiers", defaults.keyboardModifiers),
            alternateMouseButton = if (!altBtn.isNullOrEmpty() && altBtn != "null")
                runCatching { MouseButton.valueOf(altBtn) }.getOrNull()
            else null,
        )
    }

    private fun macrosToJson(macros: List<KeyboardMacroButtonConfig>): JSONArray {
        val arr = JSONArray()
        macros.forEach { macro ->
            val j = JSONObject()
            j.put("label", macro.label)
            j.put("modifiers", macro.modifiers)
            val keys = JSONArray()
            macro.keyUsages.forEach { keys.put(it) }
            j.put("keyUsages", keys)
            j.put("layoutOffsetX", macro.layoutOffsetX.toDouble())
            j.put("layoutOffsetY", macro.layoutOffsetY.toDouble())
            j.put("layoutScaleX", macro.layoutScaleX.toDouble())
            j.put("layoutScaleY", macro.layoutScaleY.toDouble())
            arr.put(j)
        }
        return arr
    }

    private fun macrosFromJson(arr: JSONArray?): List<KeyboardMacroButtonConfig> {
        if (arr == null) return emptyList()
        return List(arr.length()) { i ->
            val j = arr.getJSONObject(i)
            val keysArr = j.optJSONArray("keyUsages")
            val keys = if (keysArr != null) List(keysArr.length()) { k -> keysArr.getInt(k) } else emptyList()
            KeyboardMacroButtonConfig(
                label         = j.optString("label", ""),
                modifiers     = j.optInt("modifiers", 0),
                keyUsages     = keys,
                layoutOffsetX = j.optDouble("layoutOffsetX", 0.0).toFloat(),
                layoutOffsetY = j.optDouble("layoutOffsetY", 0.0).toFloat(),
                layoutScaleX  = j.optDouble("layoutScaleX", 1.0).toFloat(),
                layoutScaleY  = j.optDouble("layoutScaleY", 1.0).toFloat(),
            )
        }.filter { it.label.isNotBlank() && it.keyUsages.isNotEmpty() }
    }

    private inline fun <reified T : Enum<T>> enumValueOrDefault(name: String?, default: T): T =
        if (!name.isNullOrEmpty()) runCatching { enumValueOf<T>(name) }.getOrDefault(default) else default
}

File: ios/TabletHID/Services/TouchMouseConfigSerializer.swift
import Foundation

/// Converts TouchMouseConfig to and from the canonical JSON dictionary format defined in
/// spec/server-schema.md. Field names and nesting structure are identical to
/// TouchMouseConfigSerializer.kt on Android.
///
/// Note: iOS TouchMouseConfig does not yet expose subRegions or sniperEnabled/sniperDivisor
/// as first-class fields; they are preserved as pass-through values in canonical JSON where
/// available but default to empty/disabled when deserialising into the current model.
enum TouchMouseConfigSerializer {

    // MARK: - To canonical JSON

    static func toCanonicalJson(_ config: TouchMouseConfig) -> [String: Any] {
        var root: [String: Any] = [:]
        root["mode"]             = config.mode.rawValue.uppercased()
        root["sensitivity"]      = config.sensitivity
        root["scrollEnabled"]    = config.scrollEnabled
        root["invertScroll"]     = config.invertScroll
        root["sharedDynamicZone"] = config.sharedDynamicZone
        root["sharedDynamic"] = [
            "offsetX": config.sharedDynamicOffsetX,
            "offsetY": config.sharedDynamicOffsetY,
            "radius":  config.sharedDynamicRadius,
        ] as [String: Any]
        root["leftButton"]  = buttonZoneToJson(config.leftButton)
        root["rightButton"] = buttonZoneToJson(config.rightButton)
        // Sniper: iOS model stores flattened fields; serialise to nested object.
        root["sniper"] = [
            "enabled": false,
            "zone": [
                "left":   0.35,
                "top":    0.88,
                "right":  0.65,
                "bottom": 1.0,
            ],
            "divisor": 4.0,
        ] as [String: Any]
        root["macroHostDefaults"] = config.macroHostDefaults.rawValue.uppercased()
        root["macroButtons"] = config.macroButtons.map { macroToJson($0) }
        return root
    }

    // MARK: - From canonical JSON

    static func fromCanonicalJson(_ json: [String: Any]) -> TouchMouseConfig {
        let sharedDynamic = json["sharedDynamic"] as? [String: Any]
        let leftButtonJson  = json["leftButton"]  as? [String: Any]
        let rightButtonJson = json["rightButton"] as? [String: Any]
        var config = TouchMouseConfig()
        config.mode          = touchMode(from: json["mode"] as? String) ?? .touch
        config.sensitivity   = json["sensitivity"]   as? Int  ?? 5
        config.scrollEnabled = json["scrollEnabled"] as? Bool ?? true
        config.invertScroll  = json["invertScroll"]  as? Bool ?? false
        config.sharedDynamicZone = json["sharedDynamicZone"] as? Bool ?? false
        config.sharedDynamicOffsetX = (sharedDynamic?["offsetX"] as? Double) ?? 0
        config.sharedDynamicOffsetY = (sharedDynamic?["offsetY"] as? Double) ?? 0.18
        config.sharedDynamicRadius  = (sharedDynamic?["radius"]  as? Double) ?? 0.08
        config.leftButton   = buttonZoneFromJson(leftButtonJson,  defaults: config.leftButton)
        config.rightButton  = buttonZoneFromJson(rightButtonJson, defaults: config.rightButton)
        config.macroHostDefaults = macroHost(from: json["macroHostDefaults"] as? String) ?? .windows
        config.macroButtons = (json["macroButtons"] as? [[String: Any]] ?? []).compactMap { macroFromJson($0) }
        return config
    }

    // MARK: - ButtonZoneConfig helpers

    private static func buttonZoneToJson(_ btn: ButtonZoneConfig) -> [String: Any] {
        var j: [String: Any] = [:]
        j["enabled"]  = btn.enabled
        j["zoneType"] = zoneTypeName(btn.zoneType)
        j["behavior"] = clickBehaviorName(btn.behavior)
        j["staticZone"] = [
            "left":   btn.staticLeft,
            "top":    btn.staticTop,
            "right":  btn.staticRight,
            "bottom": btn.staticBottom,
        ] as [String: Any]
        j["dynamicZone"] = [
            "offsetX": btn.dynamicOffsetX,
            "offsetY": btn.dynamicOffsetY,
            "radius":  btn.dynamicRadius,
        ] as [String: Any]
        j["subRegions"] = [] as [[String: Any]]
        return j
    }

    private static func buttonZoneFromJson(_ j: [String: Any]?, defaults: ButtonZoneConfig) -> ButtonZoneConfig {
        var btn = defaults
        guard let j = j else { return btn }
        btn.enabled  = j["enabled"]  as? Bool ?? btn.enabled
        btn.zoneType = zoneType(from: j["zoneType"] as? String) ?? btn.zoneType
        btn.behavior = clickBehavior(from: j["behavior"] as? String) ?? btn.behavior
        if let staticZone = j["staticZone"] as? [String: Any] {
            btn.staticLeft   = (staticZone["left"]   as? Double) ?? btn.staticLeft
            btn.staticTop    = (staticZone["top"]    as? Double) ?? btn.staticTop
            btn.staticRight  = (staticZone["right"]  as? Double) ?? btn.staticRight
            btn.staticBottom = (staticZone["bottom"] as? Double) ?? btn.staticBottom
        }
        if let dynamicZone = j["dynamicZone"] as? [String: Any] {
            btn.dynamicOffsetX = (dynamicZone["offsetX"] as? Double) ?? btn.dynamicOffsetX
            btn.dynamicOffsetY = (dynamicZone["offsetY"] as? Double) ?? btn.dynamicOffsetY
            btn.dynamicRadius  = (dynamicZone["radius"]  as? Double) ?? btn.dynamicRadius
        }
        // subRegions: not yet supported in iOS model; skip.
        return btn
    }

    // MARK: - Macro helpers

    private static func macroToJson(_ macro: KeyboardMacroButtonConfig) -> [String: Any] {
        return [
            "label":         macro.label,
            "modifiers":     macro.modifiers,
            "keyUsages":     macro.keyUsages,
            "layoutOffsetX": macro.layoutOffsetX,
            "layoutOffsetY": macro.layoutOffsetY,
            "layoutScaleX":  macro.layoutScaleX,
            "layoutScaleY":  macro.layoutScaleY,
        ]
    }

    private static func macroFromJson(_ j: [String: Any]) -> KeyboardMacroButtonConfig? {
        let label    = j["label"] as? String ?? ""
        let keyUsages = j["keyUsages"] as? [Int] ?? []
        guard !label.isEmpty, !keyUsages.isEmpty else { return nil }
        return KeyboardMacroButtonConfig(
            label:         label,
            modifiers:     j["modifiers"]     as? Int    ?? 0,
            keyUsages:     keyUsages,
            layoutOffsetX: j["layoutOffsetX"] as? Double ?? 0,
            layoutOffsetY: j["layoutOffsetY"] as? Double ?? 0,
            layoutScaleX:  j["layoutScaleX"]  as? Double ?? 1,
            layoutScaleY:  j["layoutScaleY"]  as? Double ?? 1
        )
    }

    // MARK: - Enum helpers

    private static func touchMode(from raw: String?) -> TouchMode? {
        switch raw?.uppercased() {
        case "TOUCH": return .touch
        case "MOUSE": return .mouse
        default:      return nil
        }
    }

    private static func zoneTypeName(_ z: ZoneType) -> String {
        switch z {
        case .staticZone: return "STATIC"
        case .dynamic:    return "DYNAMIC"
        }
    }

    private static func zoneType(from raw: String?) -> ZoneType? {
        switch raw?.uppercased() {
        case "STATIC":  return .staticZone
        case "DYNAMIC": return .dynamic
        default:        return nil
        }
    }

    private static func clickBehaviorName(_ b: ClickBehavior) -> String {
        switch b {
        case .momentary: return "MOMENTARY"
        case .latching:  return "LATCHING"
        }
    }

    private static func clickBehavior(from raw: String?) -> ClickBehavior? {
        switch raw?.uppercased() {
        case "MOMENTARY": return .momentary
        case "LATCHING":  return .latching
        default:          return nil
        }
    }

    private static func macroHost(from raw: String?) -> MacroHostDefaults? {
        switch raw?.uppercased() {
        case "WINDOWS": return .windows
        case "MAC":     return .mac
        default:        return nil
        }
    }
}
